lib/pure/hashes.nim

```diff
@@ -197,8 +197,28 @@ else:
 
 proc hash*(x: float): Hash {.inline.} =
   ## Efficient hashing of floats.
-  var y = x + 0.0 # for denormalization
-  result = hash(cast[ptr Hash](addr(y))[])
+  let y = x + 0.0 # for denormalization
+  when nimvm:
+    when defined(js):
+      # workaround a JS VM bug: bug #16547
+      result = hashWangYi1(cast[int64](y))
+    else:
+      result = hashWangYi1(cast[Hash](y))
+  else:
+    when not defined(js):
+      result = hashWangYi1(cast[Hash](y))
+    else:
+      var res: Hash
+      asm """const buffer = new ArrayBuffer(8);
+    const floatBuffer = new Float64Array(buffer);
+    const uintBuffer = new BigUint64Array(buffer);
+    floatBuffer[0] = `y`;
+    `res` = uintBuffer[0];"""
+
+      # res is a `BigInt` type, but we cheat the type system
+      # and say it is a `Hash` type.
+      # TODO refactor it using bigInt once jsBigInt is ready, pending pr #1640
+      result = hashWangYi1(res)
```

Comment:
how about this to make it easier to evolve/reuse/generalize `asBigInt` in future work
```nim
when defined(js):
  proc asBigInt(x: float): int64 =
    # result is a `BigInt` type in js, but we cheat the type system
    # and say it is a `int64` type.
    # TODO refactor it using bigInt once jsBigInt is ready, pending pr #1640
    asm """
    const buffer = new ArrayBuffer(8);
    const floatBuffer = new Float64Array(buffer);
    const uintBuffer = new BigUint64Array(buffer);
    floatBuffer[0] = `x`;
    `result` = uintBuffer[0];"""

proc hash*(x: float): Hash {.inline.} =
  ## Efficient hashing of floats.
  let y = x + 0.0 # for denormalization
  when nimvm:
    when defined(js):
      # workaround a JS VM bug: bug #16547
      result = hashWangYi1(cast[int64](y))
    else:
      result = hashWangYi1(cast[Hash](y))
  else:
    when not defined(js):
      result = hashWangYi1(cast[Hash](y))
    else:
      result = hashWangYi1(asBigInt(y))
```

Reply:
I see, thanks

Reply:
```nim
block hashes:
  block hashing:
    var dummy = 0.0
    doAssert hash(dummy) == hash(-dummy)
```
can't pass this test